diff --git a/fs/exec.c b/fs/exec.c
index b1b9db5..e90e89f 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1888,11 +1888,27 @@ static int __do_execve_file(int fd, struct filename *filename,
 	return retval;
 }
 
+/* [SukiSU] Import External Function */
+#ifdef CONFIG_KSU
+extern int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
+					void *argv, void *envp, int *flags);
+#endif
+
 int do_execveat_common(int fd, struct filename *filename,
 		       struct user_arg_ptr argv,
 		       struct user_arg_ptr envp,
 		       int flags)
 {
+    /* [SukiSU] Hook execve at kernel level */
+#ifdef CONFIG_KSU
+	ksu_handle_execveat_sucompat(&fd, &filename, &argv, &envp, &flags);
+#endif
+
 	return __do_execve_file(fd, filename, argv, envp, flags, NULL);
 }
 
diff --git a/fs/open.c b/fs/open.c
index 39779f1..1dcbefb 100644
--- a/fs/open.c
+++ b/fs/open.c
@@ -466,8 +466,19 @@ static long do_faccessat(int dfd, const char __user *filename, int mode, int fla
 	return res;
 }
+
+/* [SukiSU] Import External Function */
+#ifdef CONFIG_KSU
+extern int ksu_handle_faccessat(int *dfd, const char __user **filename_user,
+			       int *mode, int *flags);
+#endif
+
 SYSCALL_DEFINE3(faccessat, int, dfd, const char __user *, filename, int, mode)
 {
+    /* [SukiSU] Hook faccessat to hide su existence */
+#ifdef CONFIG_KSU
+	int ks_flags = 0;
+	ksu_handle_faccessat(&dfd, &filename, &mode, &ks_flags);
+#endif
 	return do_faccessat(dfd, filename, mode, 0);
 }
 
diff --git a/fs/stat.c b/fs/stat.c
index 357a1db..d5087ca 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -401,6 +401,12 @@ SYSCALL_DEFINE2(newlstat, const char __user *, filename,
 	return cp_new_stat(&stat, statbuf);
 }
+
+/* [SukiSU] Import External Function */
+#ifdef CONFIG_KSU
+extern int ksu_handle_stat(int *dfd, const char __user **filename_user,
+			   int *flags);
+#endif
+
 #if !defined(__ARCH_WANT_STAT64) || defined(__ARCH_WANT_SYS_NEWFSTATAT)
 SYSCALL_DEFINE4(newfstatat, int, dfd, const char __user *, filename,
 		struct stat __user *, statbuf, int, flag)
@@ -408,6 +414,12 @@ SYSCALL_DEFINE4(newfstatat, int, dfd, const char __user *, filename,
 	struct kstat stat;
 	int error;
+    /* [SukiSU] Hook newfstatat to hide su existence */
+#ifdef CONFIG_KSU
+	ksu_handle_stat(&dfd, &filename, &flag);
+#endif
+
 	error = vfs_fstatat(dfd, filename, &stat, flag);
 	if (error)
 		return error;
 }
